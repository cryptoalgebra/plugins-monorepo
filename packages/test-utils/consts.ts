export const TEST_POOL_START_TIME = 1601906400;