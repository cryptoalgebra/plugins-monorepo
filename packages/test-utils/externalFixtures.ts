import {
  abi as FACTORY_ABI,
  bytecode as FACTORY_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/AlgebraFactory.sol/AlgebraFactory.json';
import {
  abi as TEST_CALLEE_ABI,
  bytecode as TEST_CALLEE_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/test/TestAlgebraCallee.sol/TestAlgebraCallee.json';
import {
  abi as POOL_DEPLOYER_ABI,
  bytecode as POOL_DEPLOYER_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/test/MockTimeAlgebraPoolDeployer.sol/MockTimeAlgebraPoolDeployer.json';
import {
  abi as POOL_ABI,
  bytecode as POOL_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/test/MockTimeAlgebraPool.sol/MockTimeAlgebraPool.json';
import {
  abi as COM_VAULT_DEPLOYER_ABI,
  bytecode as COM_VAULT_DEPLOYER_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/AlgebraCommunityVault.sol/AlgebraCommunityVault.json';
import {
  abi as COM_VAULT_STUB_DEPLOYER_ABI,
  bytecode as COM_VAULT_STUB_DEPLOYER_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/AlgebraVaultFactoryStub.sol/AlgebraVaultFactoryStub.json';
import {
  abi as ENTRYPOINT_ABI,
  bytecode as ENTRYPOINT_BYTECODE,
} from '@cryptoalgebra/integral-periphery/artifacts/contracts/AlgebraCustomPoolEntryPoint.sol/AlgebraCustomPoolEntryPoint.json';
import {
  abi as POOL_DEPLOYER_CORE_ABI,
  bytecode as POOL_DEPLOYER_CORE_BYTECODE,
} from '@cryptoalgebra/integral-core/artifacts/contracts/AlgebraPoolDeployer.sol/AlgebraPoolDeployer.json';

import { ethers } from 'hardhat';
import {
  MockTimeAlgebraPoolDeployer,
  AlgebraPoolDeployer,
  AlgebraFactory,
  AlgebraCommunityVault,
  TestAlgebraCallee,
  MockTimeAlgebraPool,
  TestERC20,
} from '@cryptoalgebra/integral-core/typechain';

import {
  AlgebraCustomPoolEntryPoint
} from '@cryptoalgebra/integral-periphery/typechain';
import { getCreateAddress } from 'ethers';

interface TokensFixture {
  token0: TestERC20;
  token1: TestERC20;
}

export async function tokensFixture(): Promise<TokensFixture> {
  const tokenFactory = await ethers.getContractFactory('TestERC20');
  const tokenA = (await tokenFactory.deploy(2n ** 255n)) as any as TestERC20 & { address: string };
  const tokenB = (await tokenFactory.deploy(2n ** 255n)) as any as TestERC20 & { address: string };

  tokenA.address = await tokenA.getAddress();
  tokenB.address = await tokenB.getAddress();

  const [token0, token1] = [tokenA, tokenB].sort((_tokenA, _tokenB) => (_tokenA.address.toLowerCase() < _tokenB.address.toLowerCase() ? -1 : 1));

  return { token0, token1 };
}

interface MockPoolDeployerFixture extends TokensFixture {
  poolDeployer: MockTimeAlgebraPoolDeployer;
  swapTargetCallee: TestAlgebraCallee;
  factory: AlgebraFactory;
  createPool(firstToken?: TestERC20, secondToken?: TestERC20): Promise<MockTimeAlgebraPool>;
}
export const algebraPoolDeployerMockFixture: () => Promise<MockPoolDeployerFixture> = async () => {
  const { token0, token1 } = await tokensFixture();

  const [deployer] = await ethers.getSigners();
  // precompute
  const poolDeployerAddress = getCreateAddress({
    from: deployer.address,
    nonce: (await ethers.provider.getTransactionCount(deployer.address)) + 1,
  });

  const factoryFactory = await ethers.getContractFactory(FACTORY_ABI, FACTORY_BYTECODE);
  const factory = (await factoryFactory.deploy(poolDeployerAddress)) as any as AlgebraFactory;

  const poolDeployerFactory = await ethers.getContractFactory(POOL_DEPLOYER_ABI, POOL_DEPLOYER_BYTECODE);
  const poolDeployer = (await poolDeployerFactory.deploy()) as any as MockTimeAlgebraPoolDeployer;

  const ADMIN_ROLE = await factory.POOLS_ADMINISTRATOR_ROLE();
  await factory.grantRole(ADMIN_ROLE, poolDeployer);

  const vaultFactory = await ethers.getContractFactory(COM_VAULT_DEPLOYER_ABI, COM_VAULT_DEPLOYER_BYTECODE);
  const vault = (await vaultFactory.deploy(factory, deployer.address)) as any as AlgebraCommunityVault;

  const vaultFactoryStubFactory = await ethers.getContractFactory(COM_VAULT_STUB_DEPLOYER_ABI, COM_VAULT_STUB_DEPLOYER_BYTECODE);
  const vaultFactoryStub = await vaultFactoryStubFactory.deploy(vault);

  await factory.setVaultFactory(vaultFactoryStub);

  const calleeContractFactory = await ethers.getContractFactory(TEST_CALLEE_ABI, TEST_CALLEE_BYTECODE);
  const swapTargetCallee = (await calleeContractFactory.deploy()) as any as TestAlgebraCallee;

  const MockTimeAlgebraPoolFactory = await ethers.getContractFactory(POOL_ABI, POOL_BYTECODE);

  return {
    poolDeployer,
    swapTargetCallee,
    token0,
    token1,
    factory,
    createPool: async (firstToken = token0, secondToken = token1) => {
      await poolDeployer.deployMock(factory, firstToken, secondToken);

      const sortedTokens =
        BigInt(await firstToken.getAddress()) < BigInt(await secondToken.getAddress())
          ? [await firstToken.getAddress(), await secondToken.getAddress()]
          : [await secondToken.getAddress(), await firstToken.getAddress()];
      const poolAddress = await poolDeployer.computeAddress(sortedTokens[0], sortedTokens[1]);
      return MockTimeAlgebraPoolFactory.attach(poolAddress) as any as MockTimeAlgebraPool;
    },
  };
};



interface algebraDeployFixture extends TokensFixture {
  poolDeployer: AlgebraPoolDeployer;
  factory: AlgebraFactory;
  entryPoint: AlgebraCustomPoolEntryPoint;
}
export const algebraCoreFixture: () => Promise<algebraDeployFixture> = async () => {
  const { token0, token1 } = await tokensFixture();

  const [deployer] = await ethers.getSigners();
  // precompute
  const poolDeployerAddress = getCreateAddress({
    from: deployer.address,
    nonce: (await ethers.provider.getTransactionCount(deployer.address)) + 1,
  });

  const factoryFactory = await ethers.getContractFactory(FACTORY_ABI, FACTORY_BYTECODE);
  const factory = (await factoryFactory.deploy(poolDeployerAddress)) as any as AlgebraFactory;

  const poolDeployerFactory = await ethers.getContractFactory(POOL_DEPLOYER_CORE_ABI, POOL_DEPLOYER_CORE_BYTECODE);
  const poolDeployer = (await poolDeployerFactory.deploy(factory)) as any as AlgebraPoolDeployer;

  const ADMIN_ROLE = await factory.POOLS_ADMINISTRATOR_ROLE();
  await factory.grantRole(ADMIN_ROLE, poolDeployer);

  const entrypointFactory = await ethers.getContractFactory(ENTRYPOINT_ABI, ENTRYPOINT_BYTECODE);
  const entryPoint = await entrypointFactory.deploy(factory) as any as AlgebraCustomPoolEntryPoint;
  
  let customPoolDeployerRole = await factory.CUSTOM_POOL_DEPLOYER()
  let poolAdministratorRole = await factory.POOLS_ADMINISTRATOR_ROLE()
  await factory.grantRole(customPoolDeployerRole, await entryPoint.getAddress());
  await factory.grantRole(poolAdministratorRole, await entryPoint.getAddress());

  return {
    poolDeployer,
    token0,
    token1,
    factory,
    entryPoint
  };
};
