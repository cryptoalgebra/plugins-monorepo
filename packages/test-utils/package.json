{
  "name": "test-utils",
  "version": "1.0.0",
  "private": "true",
  "description": "",
  "devDependencies": {
    "@ethersproject/abstract-provider": "^5.8.0",
    "bignumber.js": "^9.2.1",
    "mocha-chai-jest-snapshot": "^1.1.6"
  }
}
