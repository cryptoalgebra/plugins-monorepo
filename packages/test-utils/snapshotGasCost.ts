import { TransactionReceipt, TransactionResponse } from '@ethersproject/abstract-provider';
import { expect } from './expect';
import { Contract, ContractTransactionResponse } from 'ethers';

export default async function snapshotGasCost(
  x:
    | TransactionResponse
    | Promise<TransactionResponse>
    | ContractTransactionResponse
    | Promise<ContractTransactionResponse>
    | TransactionReceipt
    | Promise<BigInt>
    | BigInt
    | Promise<bigint>
    | bigint
    | Contract
    | Promise<Contract>
): Promise<void> {
  const resolved = await x;
  if (typeof resolved == 'bigint' || resolved instanceof BigInt || resolved instanceof Number) {
    expect(Number(resolved)).toMatchSnapshot();
  } else if ('deploymentTransaction' in resolved) {
    const receipt = await resolved.deploymentTransaction()?.wait();
    if (receipt) {
      expect(Number(receipt.gasUsed)).toMatchSnapshot();
    } else {
        console.warn('⚠️Deployment transaction returned null. Skipping gas usage snapshot.\n', new Error().stack);
    }
  } else if ('wait' in resolved) {
    const waited = await resolved.wait();
    if (waited) {
        expect(Number(waited.gasUsed)).toMatchSnapshot();
    } else {
        console.warn('⚠️Transaction returned null. Skipping gas usage snapshot.\n', new Error().stack);
    }
  }
}
