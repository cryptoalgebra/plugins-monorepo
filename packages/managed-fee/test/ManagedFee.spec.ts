import { expect } from 'test-utils/expect';
import { ethers } from 'hardhat';
import { Wallet, AbiCoder, keccak256 } from 'ethers'
import { ManagedSwapFeeTest } from '../typechain';
import { loadFixture } from '@nomicfoundation/hardhat-toolbox/network-helpers';
import snapshotGasCost from 'test-utils/snapshotGasCost';

describe('ManagedSwapFee', () => {
  let wallet: Wallet, other: Wallet;
  let managedSwapFeePlugin: ManagedSwapFeeTest;
  let pluginData: string;
  let nonce: string;
  let fee: number;
  let user: string;
  let expireTime: number;

  async function managedSwapFeeFixture() {
    const factory = await ethers.getContractFactory('ManagedSwapFeeTest');
    return (await factory.deploy()) as any as ManagedSwapFeeTest;
  }

  async function generatePluginData(nonce: string, fee: number, user: string, expireTime: number, signer: Wallet): Promise<string> {
    let hash = keccak256(AbiCoder.defaultAbiCoder().encode(
      ['bytes32', 'uint24', 'address', 'uint32'],
      [nonce, fee, user, expireTime])
    );

    const hashBytes = Buffer.from(hash.slice(2), 'hex');
    let signature = await signer.signMessage(hashBytes);

    return AbiCoder.defaultAbiCoder().encode(
      ['tuple(bytes32, uint24, address, uint32, bytes)'],
      [[nonce, fee, user, expireTime, signature]]
    );
  }

  beforeEach('deploy ManagedSwapFeeTest', async () => {
    [wallet, other] = await (ethers as any).getSigners();
    managedSwapFeePlugin = await loadFixture(managedSwapFeeFixture);
  });

  describe('#getManagedFee', () => {
    beforeEach('set config', async () => {
      let provider = ethers.provider
      const block = await provider.getBlock('latest');

      nonce ="0x0000000000000000000000000000000000000000000000000000000000000001"
      fee = 1000
      user = wallet.address
      expireTime = block!.timestamp + 1000

      pluginData = await generatePluginData(nonce, fee, user, expireTime, wallet)
      await managedSwapFeePlugin.setWhitelistStatus(wallet.address, true);

    });

    it('fee is used on swap', async () => {
      await expect(managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)).to.be.emit(managedSwapFeePlugin, "Fee").withArgs(1000)
    });

    it('revert if signer is not whitelisted', async () => {
      pluginData = await generatePluginData(nonce, fee, user, expireTime, other)
      await expect(managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)).to.be.revertedWithCustomError(managedSwapFeePlugin,"NotWhitelisted")
    });

    it('nonce can not be used twice', async () => {
      await managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)
      await expect(managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)).to.be.revertedWithCustomError(managedSwapFeePlugin,"InvalidNonce")
    });

    it('fee can not exceed 100%', async () => {
      fee = 1000000 // 100%
      pluginData = await generatePluginData(nonce, fee, user, expireTime, wallet)
      await expect(managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)).to.be.revertedWithCustomError(managedSwapFeePlugin,"FeeExceedsLimit")
    });

    it('revert if signature expired', async () => {
      let provider = ethers.provider
      const block = await provider.getBlock('latest');

      expireTime = block!.timestamp - 1
      pluginData = await generatePluginData(nonce, fee, user, expireTime, wallet)
      await expect(managedSwapFeePlugin.connect(wallet).getFeeForSwap(pluginData)).to.be.revertedWithCustomError(managedSwapFeePlugin,"Expired")
    });

    it('revert if user and caller missmatch', async () => {
      await expect(managedSwapFeePlugin.connect(other).getFeeForSwap(pluginData)).to.be.revertedWithCustomError(managedSwapFeePlugin,"NotAllowed")
    });

    it('gas cost', async () => {
      await snapshotGasCost(managedSwapFeePlugin.getGasCostOfGetFeeForSwap(pluginData));
    });

  });

});