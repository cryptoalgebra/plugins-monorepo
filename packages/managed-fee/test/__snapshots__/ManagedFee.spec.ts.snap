// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagedSwapFee #getManagedFee gas cost 1`] = `54780`;
